modules/odm_extract_utm/src/UtmExtractor.cpp

```diff
@@ -273,81 +280,54 @@ bool UtmExtractor::convert(const double &lon, const double &lat, const double &a
-           throw UtmExtractorException("Image is missing GPS Altitude data");
-       }
-    }
+  Exiv2::Exifdatum& latitudeTag = exifData["Exif.GPSInfo.GPSLatitude"];
+  Exiv2::Exifdatum& latitudeRef = exifData["Exif.GPSInfo.GPSLatitudeRef"];
+  Exiv2::Exifdatum& longitudeTag = exifData["Exif.GPSInfo.GPSLongitude"];
+  Exiv2::Exifdatum& longitudeRef = exifData["Exif.GPSInfo.GPSLongitudeRef"];
+  Exiv2::Exifdatum& altitudeTag = exifData["Exif.GPSInfo.GPSAltitude"];
+  Exiv2::Exifdatum& altitudeRef = exifData["Exif.GPSInfo.GPSAltitudeRef"];
+
+  // Latitude: parse into a double
+  if (latitudeTag.count() < 3)
+    throw UtmExtractorException("Image is missing GPS Latitude data");
+  else {
+    Exiv2::URational rLat[] = {latitudeTag.toRational(0), latitudeTag.toRational(1), latitudeTag.toRational(2)};
+    bool south = (strcmp(latitudeRef.toString().c_str(), "S") == 0);
+    double degrees, minutes, seconds;
+
+    degrees = (double)rLat[0].first / (double)rLat[0].second;
+    minutes = (double)rLat[1].first / (double)rLat[1].second / 60.0;
+    seconds = (double)rLat[2].first / (double)rLat[2].second / 3600.0;
+    lat = (south ? -1 : 1) * (degrees + minutes + seconds);
   }
 
-  if (lonStr.empty() || latStr.empty()) {
-    throw UtmExtractorException("No valid GPS position found");
-    return false;
+  // Longitude
+  if (longitudeTag.count() < 3)
+    throw UtmExtractorException("Image is missing GPS Longitude data");
+  else {
+    Exiv2::URational rLon[] = {longitudeTag.toRational(0), longitudeTag.toRational(1), longitudeTag.toRational(2)};
+    bool west = (strcmp(longitudeRef.toString().c_str(), "W") == 0);
+    double degrees, minutes, seconds;
+
+    degrees = (double)rLon[0].first / (double)rLon[0].second;
+    minutes = (double)rLon[1].first / (double)rLon[1].second / 60.0;
+    seconds = (double)rLon[2].first / (double)rLon[2].second / 3600.0;
+    lon = (west ? -1 : 1) * (degrees + minutes + seconds);
   }
 
-  // Parse longitude
-  std::string hemisphere;
-  double degrees, minutes, seconds;
-  std::istringstream istr(lonStr);
-  char degChar = 'd', minChar = 'm', secChar = 's';
-  istr >> hemisphere >> degrees >> degChar >> minutes >> minChar >> seconds >> secChar;
-  lon = (hemisphere == "W" ? -1 : 1) * (degrees + minutes/60.0 + seconds/3600.0);
-
-  // Parse latitude
-  istr.clear();
-  istr.str(latStr);
-  istr >> hemisphere >> degrees >> degChar >> minutes >> minChar >> seconds >> secChar;
-  lat = (hemisphere == "S" ? -1 : 1) * (degrees + minutes/60.0 + seconds/3600.0);
-
-  if (!altStr.empty())
-  {
-    char meterUnitChar;
-    istr.clear();
-    istr.str(altStr);
-    istr >> alt >> meterUnitChar;
+  // Altitude
+  if (altitudeTag.count() < 1)
+    throw UtmExtractorException("Image is missing GPS Altitude data");
+  else {
+    Exiv2::URational rAlt = altitudeTag.toRational(0);
+    bool below = (altitudeRef.count() >= 1 && altitudeRef.toLong() == 1);
+    alt = (below ? -1 : 1) * (double)rAlt.first / (double)rAlt.second;
```

Comment:
Could rAlt.second ever be 0?

Reply:
I hope not, it's being pulled from a rational  type: http://exiv2.org/tags.html


Tag (hex) | Tag (dec) | IFD | Key | Type | Tag description
-- | -- | -- | -- | -- | --
0x0006 | 6 | GPSInfo | Exif.GPSInfo.GPSAltitude | Rational | Indicates the altitude based on the reference in GPSAltitudeRef. Altitude is expressed as one RATIONAL value. The reference unit is meters.